zul/src/archive/web/js/zul/inp/InputWidget.js

```diff
@@ -678,13 +678,29 @@ zul.inp.InputWidget = zk.$extends(zul.Widget, {
 		}
 	},
 	_shallIgnore: function (evt, keys) {
+		function escapeRegExp(text) {
+			return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
+		}
+		if (evt.keyCode || evt.charCode) {
 		// ZK-1736 add metakey on mac
-		if (zk.mac && evt.metaKey)
-			return;
-		else {
-			var code = (zk.ie < 11 || zk.opera) ? evt.keyCode : evt.charCode;
-			if (!evt.altKey && !evt.ctrlKey && _keyIgnorable(code)
-			&& keys.indexOf(String.fromCharCode(code)) < 0) {
+			if (zk.mac && evt.metaKey)
+				return;
+			else {
+				var code = (zk.ie < 11 || zk.opera) ? evt.keyCode : evt.charCode;
+				if (!evt.altKey && !evt.ctrlKey && _keyIgnorable(code)
+				&& keys.indexOf(String.fromCharCode(code)) < 0) {
+					evt.stop();
+					return true;
+				}
+			}
+		} else {
+			var orgEvtText = evt.domEvent.originalEvent.clipboardData.getData('text'),
+				text = (orgEvtText) ? orgEvtText :
+				window.clipboardData.getData('text'),
+				minusIndex = keys.indexOf('-');
+			keys = escapeRegExp(keys);
+			var regKey = '^[' + keys + ']+$';
```

Comment:
The method is not widely used, so just inline it
```js
var regKey = '^[' + keys.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&') + ']+$';
```

Reply:
ok